### PR TITLE
Add support for Slice/Array of structs

### PR DESCRIPTION
**Pull request by THE108**
Changes:
- Added support for Slice/Array of structs
- Removed unused params

### CODE REVIEW

File: mold_test.go [hunk 1]
```diff
@@ -601,3 +601,72 @@ func TestDiveKeys(t *testing.T) {
...
+	})
+
+	var tt Test
+
+	err := set.Struct(context.Background(), &tt)
+	Equal(t, err, nil)
+	Equal(t, len(tt.Arr), 2)
+	Equal(t, len(tt.ArrDive), 2)
+	Equal(t, tt.Arr[0].String, "default")
+	Equal(t, tt.Arr[1].String, "default")
```
Comment (deankarn) [inline mold_test.go]: The above two lines should not be `default` but blank instead.

File: mold_test.go [hunk 2]
```diff
@@ -601,3 +601,72 @@ func TestDiveKeys(t *testing.T) {
...
+	Equal(t, tt.Inner.String, "default")
+
+	tt2 := Test{
+		Arr: make([]InnerStruct, 1),
+	}
+
+	err = set.Struct(context.Background(), &tt2)
+	Equal(t, err, nil)
+	Equal(t, len(tt2.Arr), 1)
+	Equal(t, tt2.Arr[0].String, "default")
```
Comment (deankarn) [inline mold_test.go]: This should be blank instead of `default`

File: mold_test.go [hunk 3]
```diff
@@ -601,3 +601,72 @@ func TestDiveKeys(t *testing.T) {
...
+
+	tt3 := Test{
+		Arr: []InnerStruct{{"ok"}},
+	}
+
+	err = set.Struct(context.Background(), &tt3)
+	NotEqual(t, err, nil)
```
Comment (deankarn) [inline mold_test.go]: This should be `nil`

File: mold_test.go [hunk 4]
```diff
@@ -601,3 +601,72 @@ func TestDiveKeys(t *testing.T) {
...
+	Equal(t, err.Error(), "ALREADY OK")
+
+	tt4 := Test{
+		ArrNoTag: make([]InnerStruct, 1),
+	}
+
+	err = set.Struct(context.Background(), &tt4)
+	Equal(t, err, nil)
+	Equal(t, len(tt4.ArrNoTag), 1)
+	Equal(t, tt4.ArrNoTag[0].String, "default")
```
Comment (deankarn) [inline mold_test.go]: This should be blank.

**Comment by deankarn** [comment]
@THE108 this PR, based off of yours, should fix the issue https://github.com/go-playground/mold/pull/7

Just wanted to run by you first to ensure it solves your use case before merging :)

**Comment by THE108** [comment]
Hey @deankarn

> Just wanted to run by you first to ensure it solves your use case before merging :)

Actually, it does not.

In my PR I intended to support both cases:
- Slice/Array of structs has `dive` tag
- Slice/Array of structs doesn't have any `mold` related tags

The second case opposes this point of yours  - `checking for the array/slice at the end of setByField isn't the right solution.`

Your solution does fix the issue but I think it would be better from the usability perspective to dive into Slice/Array **of structs** with no additional tag needed as you dive into struct field (field of type struct) now.

**Comment by deankarn** [comment]
@THE108 The way this package is intended to work is that `mold` related actions can apply at the struct level or the field level. To get to the field level you must traverse the struct and why structs are automatically traversed and this should be the only automatic traversal, all other actions should be explicit.

>Slice/Array of structs doesn't have any mold related tags

The above would violate how this package is intended to work and would force users into a specific behaviour. eg. (using your test example)

```go
        type InnerStruct struct {
		String string `s:"defaultStr"`
	}

	type Test struct {
		Inner    InnerStruct
		Arr      []InnerStruct `s:"defaultArr"`
		ArrDive  []InnerStruct `s:"defaultArr,dive"`
		ArrNoTag []InnerStruct
	}
```

The way that I have fixed this it allows someone to set a default array in the `Arr` field without having to traverse the array nor set the default on the inner struct if desired. This gives added flexibility and gives a distinction between `s:"defaultArr"` and `s:"defaultArr,dive"` which your solution did not.

Another reason the diving should be explicit is that the library does not know when it should. eg.
take the tags `s:"defaultArr,othertag1,othertag2"`, in this case when should the library dive? before all the tags are applied? after all tags? inbetween tags?
This is not a decision the library should be making and why it needs to remain explicit.

Lastly traversing slices automatically could be a very heavy operation. in your case it was a small number of items, but it could just as easily have been million of items that someone might not want or need to traverse eg. a million integers. Yes someone could use the `-` ignore tag however this would cause this libary to be implicit. eg.
A new developer comes along and ads some new slice field of structs to the struct being passed to `mold` but doesn't add `-`, the code would now be traversing without this developers knowledge. This is something that should be avoided and why I handle it at the library level to ensure the developer workflow is smoother.

I'm sorry this change doesn't 100% suite your use case goals of automatic traversal of arrays/slices, however, this would have changed the library in a way I was not comfortable with.

The change does fix the original issue and allows for your desired functionality, just having to add the `dive` tag explicitly. I hope this explains why automatic traversal of arrays/slices can be sub-optimal.

**Pull request closed without merging**